Quantum-Control-Applications/Superconducting/Two-Flux-Tunable-Transmons/Use Case 4 - Single- and Two-Qubit State and Process Tomography/single-qubit-state-tomography.py

```diff
@@ -0,0 +1,181 @@
+    - Having calibrated the readout (readout_frequency, amplitude, duration_optimization IQ_blobs) for better SNR, and having
+      saved the derived readout threshold value in the config
+    - Set the desired flux bias in the case of flux-tunable qubits
+
+This script implements the logic outlined in the notebook
+https://github.com/bornman-nick/quantum-state-and-process-tomography. See that
+notebook for a detailed explanation of standard single qubit state
+tomography.
+"""
+
+from qm.qua import *
+from qm import QuantumMachinesManager
+from qm import SimulationConfig
+from configuration import *
+from qualang_tools.results import progress_counter, fetching_tool
+import numpy as np
+
+
+###################
+# The QUA program #
+###################
+
+# qubit under test, assuming there are multiple qubits
+# with XY line elements "q<qubit>_xy", flux lines
+# "q<qubit>_z", and readout resonator elements "rr<qubit>"
+qubit = 1
+
+if qubit == 1:
+    threshold = ge_threshold_q1
+elif qubit == 2:
+    threshold = ge_threshold_q2
+else:
+    raise ValueError(f"Incorrect qubit number chosen")
+
+
+n_avg = 10_000
+
+
+# subroutine to prepare desired qubit state
+def prepare_state(qubit):
+    # write whatever QUA code you need in order to create the
+    # state to perform tomography on, from an initial
+    # ground state. For example, to create the |1> state
+    play("y180", f"q{qubit}_xy")
+
+
+with program() as single_qubit_state_tomography:
+    n = declare(int)  # QUA variable for average loop
+    n_st = declare_stream()  # Stream for the averaging iteration 'n'
+
+    state = declare(bool)  # QUA variable for the qubit state
+    state_st = declare_stream()  # Stream for the qubit state
+
+    p = declare(int)  # QUA variable for switching between projections
+
+    I = declare(fixed)
+    Q = declare(fixed)
+
+    with for_(n, 0, n < n_avg, n + 1):  # QUA for_ loop for averaging
+        with for_(p, 0, p <= 2, p + 1):  # QUA for_ loop for switching between basis changes
+
+            prepare_state(qubit)
+            align()
+
+            with switch_(c):
+                with case_(0):  # basis X
+
+                    # Map the X-component of the Bloch vector onto the Z-axis
+                    # 1/sqrt(2)(|0>+|1>) -> |0>; 1/sqrt(2)(|0>-|1>) -> |1>
+                    play("-y90", f"q{qubit}_xy")
+
+                    align(f"q{qubit}_xy", f"rr{qubit}")
+
+                with case_(1):  # basis Y
+
+                    # Map the Y-component of the Bloch vector onto the Z-axis
+                    # 1/sqrt(2)(|0>+i|1>) -> |0>; 1/sqrt(2)(|0>-i|1>) -> |1>
+                    play("x90", f"q{qubit}_xy")
+
+                    align(f"q{qubit}_xy", f"rr{qubit}")
+
+                with case_(2):  # basis Z
+
+                    # TODO: SHOULD WE ADD A wait STATEMENT
+                    # WITH TIME EQUAL TO THE TIME OF AN X90 OR -Y90 PULSE?
```

Comment:
I don't think so. @KevinAVR  what do you think?

Reply:
I don't think it necessary. Just a thought I had at the time, but it's probably not necessary.